DigitalLearningSolutions.Data.Tests/Services/LoginServiceTests.cs

```diff
@@ -1,239 +1,461 @@
+                .Returns((new List<AdminUser> { adminUser }, new List<DelegateUser> { delegateUser }));
+            A.CallTo(() => userVerificationService.VerifyUsers(Password, A<List<AdminUser>>._, A<List<DelegateUser>>._))
+                .Returns(
+                    new UserAccountSet(new List<AdminUser> { adminUser }, new List<DelegateUser> { delegateUser })
+                );
+            A.CallTo(
+                () => userVerificationService.GetVerifiedAdminUserAssociatedWithDelegateUser(
+                    delegateUser,
+                    Password
+                )
+            ).Returns(null);
+            A.CallTo(
+                () => userVerificationService.GetVerifiedDelegateUsersAssociatedWithAdminUser(
+                    null,
+                    Password
+                )
+            ).Returns(new List<DelegateUser>());
+            A.CallTo(() => userService.GetUsersWithActiveCentres(null, A<List<DelegateUser>>._))
+                .Returns((null, new List<DelegateUser> { delegateUser }));
+            A.CallTo(() => userService.GetUserCentres(null, A<List<DelegateUser>>._)).Returns(
+                new List<CentreUserDetails>
+                    { new CentreUserDetails(delegateUser.CentreId, delegateUser.CentreName, true) }
+            );
 
             // When
-            var returnedAdminUser = loginService.GetVerifiedAdminUserAssociatedWithDelegateUser(
-                UserTestHelper.GetDefaultDelegateUser(),
-                "password"
-            );
+            var result = loginService.AttemptLogin(Username, Password);
 
             // Then
-            Assert.IsNull(returnedAdminUser);
+            using (new AssertionScope())
+            {
+                A.CallTo(() => userService.IncrementFailedLoginCount(adminUser)).MustNotHaveHappened();
+                result.LoginAttemptResult.Should().Be(LoginAttemptResult.LogIntoSingleCentre);
+                result.LogInAdmin.Should().BeNull();
+                result.LogInDelegates.Single().Should().Be(delegateUser);
+            }
         }
 
         [Test]
         public void
-            GetVerifiedAdminUserAssociatedWithDelegateUser_Returns_nothing_when_associated_admin_account_is_at_different_centre()
+            AttemptLogin_find_multiple_delegates_and_returns_choose_a_centre_login_result()
         {
             // Given
-            var associatedAdminUser = UserTestHelper.GetDefaultAdminUser(centreId: 5);
-            A.CallTo(() => userDataService.GetAdminUserByUsername(A<string>._)).Returns(associatedAdminUser);
-            A.CallTo(() => cryptoService.VerifyHashedPassword(A<string>._, A<string>._)).Returns(true);
+            var delegateUser = UserTestHelper.GetDefaultDelegateUser();
+            var secondDelegateUser = UserTestHelper.GetDefaultDelegateUser(2, 3);
+            var delegateUsers = new List<DelegateUser> { delegateUser, secondDelegateUser };
+            A.CallTo(() => userService.GetUsersByUsername(Username))

```

Comment:
The further we get through the AttemptLogin function, the more CallTo(...) we have to set up. Some of these can make the tests feel a bit cluttered and hard to read. How feasible would it be to make one or more private methods such as "GivenSuccessfulLoginUpToInitialPasswordVerification" that set up a bunch of calls necessary to reach that specific step (e.g. Initial password verification where VerifyUsers is called) by taking in the expected AdminUsers and DelegateUsers and setting up all the necessary CallTo()? Might be able to make several that use each other for multiple stages, and should help squash the tests down a bit.